### PR TITLE
fix error in npol_to_mzp transformation

### PR DESCRIPTION
**Pull request by s0larish**
## PR summary

*What does this PR introduce?*

Modifies npol_to_mzp in transforms.py by inverting Eq 44 in DeForest et al (2022)

### CODE REVIEW

File: solpolpy/transforms.py [hunk 1]
```diff
@@ -62,39 +62,40 @@ def npol_to_mzp(input_collection, offset_angle=0*u.degree, **kwargs):
...
+                             for j in range(3)] for i in range(3)])
+
+    for i, key in enumerate(key for key in input_keys if key != 'alpha'):
+        data_npol[:, :, i, 0] = input_collection[key].data
+
+    try:
+        conv_matrix_inv = np.linalg.inv(conv_matrix)
+    except np.linalg.LinAlgError as err:
+        if "Singular matrix" in str(err):
+            raise ValueError("Conversion matrix is degenerate")
```
Comment (jmbhughes) [inline solpolpy/transforms.py:83]: Can we make this a custom error?

File: solpolpy/transforms.py [hunk 2]
```diff
@@ -62,39 +62,40 @@ def npol_to_mzp(input_collection, offset_angle=0*u.degree, **kwargs):
     Notes
     ------
     Equation 44 in DeForest et al. 2022.
-
-    """""
+    """
     input_keys = list(input_collection.keys())
-    input_dict = {}
-    in_list = list(input_collection)
+    phi = [input_collection[key].meta['POLAR'] for key in input_keys if key != 'alpha']*u.degree
```
Comment (jmbhughes) [inline solpolpy/transforms.py:67]: I know phi comes from the paper, but it's a bit of an opaque name.

File: solpolpy/transforms.py [hunk 3]
```diff
@@ -62,39 +62,40 @@ def npol_to_mzp(input_collection, offset_angle=0*u.degree, **kwargs):
...
+    data_mzp_solar = np.matmul(conv_matrix_inv, data_npol)
+
+    meta = copy.copy(input_collection[input_keys[0]].meta)
+    metas = [{**meta, 'POLAR': angle} for angle in [-60, 0, 60] * u.degree]
     mask = combine_all_collection_masks(input_collection)
-    Bmzp_cube = [("M", NDCube(Bmzp[-60 * u.degree], wcs=input_collection[input_keys[0]].wcs, mask=mask, meta=metaM)),
-                 ("Z", NDCube(Bmzp[0 * u.degree], wcs=input_collection[input_keys[0]].wcs, mask=mask, meta=metaZ)),
-                 ("P", NDCube(Bmzp[60 * u.degree], wcs=input_collection[input_keys[0]].wcs, mask=mask, meta=metaP))]
-    for p_angle in in_list:
+    Bmzp_cube = [(key, NDCube(data_mzp_solar[:, :, i, 0], wcs=input_collection[input_keys[0]].wcs,
```
Comment (jmbhughes) [inline solpolpy/transforms.py:92]: I don't really like calling this a cube. It's a list of cubes.

**Review by jmbhughes — CHANGES_REQUESTED** [review]
See my comments please. Overall, can you describe why this is an improvement over the existing function?

File: solpolpy/transforms.py [hunk 4]
```diff
@@ -62,39 +62,40 @@ def npol_to_mzp(input_collection, offset_angle=0*u.degree, **kwargs):
...
+    for i, key in enumerate(key for key in input_keys if key != 'alpha'):
+        data_npol[:, :, i, 0] = input_collection[key].data
+
+    try:
+        conv_matrix_inv = np.linalg.inv(conv_matrix)
+    except np.linalg.LinAlgError as err:
+        if "Singular matrix" in str(err):
+            raise ValueError("Conversion matrix is degenerate")
+        else:
+            raise
```
[the remark this answers is not in the document]
Comment (s0larish) [reply solpolpy/transforms.py:85]: If the conversion matrix turns out to be singular and inversion could not be done

[the remark this answers is not in the document]
Comment (jmbhughes) [reply solpolpy/transforms.py:85]: Yes. But you just have `else: raise` without anything actually being raised.

[the remark this answers is not in the document]
Comment (s0larish) [reply solpolpy/transforms.py]: removed the else condition in the new push

**Pull request merged**